### PR TITLE
Reimplement APM UDP Socket methods on top of task variants and consolidate related tests

### PR DESCRIPTION
**Pull request by antonfirsov**
Contributes to #43845.

- Reimplements `(Begin|End)(SendTo|ReceiveFrom|ReceiveMessageFrom)` on top of the new Task-based variants, and deletes unused code
- Moves all related argument validation tests from `ArgumentValidationTests.cs` to `SendTo.cs`, `ReceiveFrom.cs` and `ReceiveMessageFrom.cs`, and extends coverage. Note that the new test code does not validate `ArgumentException.ParamName`,  since it's very inconsistent across all the overloads.

@geoffkizer PTAL.

### CODE REVIEW

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs [hunk 1]
```diff
@@ -2652,190 +2428,33 @@ public int EndReceiveMessageFrom(IAsyncResult asyncResult, ref SocketFlags socke
...
-                _rightEndPoint = oldEndPoint;
-                _localEndPoint = null;
-                throw;
-            }
+            ValidateReceiveFromEndpointAndState(remoteEP, nameof(remoteEP));
 
-            // Throw an appropriate SocketException if the native call fails synchronously.
-            UpdateReceiveSocketErrorForDisposed(ref errorCode, bytesTransferred: 0);
-            if (!CheckErrorAndUpdateStatus(errorCode))
+            Task<SocketReceiveFromResult> t = ReceiveFromAsync(new Memory<byte>(buffer, offset, size), socketFlags, remoteEP).AsTask();
```
Comment (gfoidl) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:2448]:
```suggestion
            Task<SocketReceiveFromResult> t = ReceiveFromAsync(buffer.AsMemory(offset, size), socketFlags, remoteEP).AsTask();
```

**Review by gfoidl — COMMENTED** [review]
For creating the Memories be consistent and always use the extension method?

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs [hunk 2]
```diff
@@ -2546,97 +2404,15 @@ public IAsyncResult BeginReceiveMessageFrom(byte[] buffer, int offset, int size,
...
-            // WSARecvMsg; all that matters is that we generate a unique-to-this-call SocketAddress
-            // with the right address family
-            Internals.SocketAddress socketAddress = Serialize(ref remoteEP);
-
-            // Guarantee to call CheckAsyncCallOverlappedResult if we call SetUnamangedStructures with a cache in order to
-            // avoid a Socket leak in case of error.
-            SocketError errorCode = SocketError.SocketError;
-            try
+            Task<SocketReceiveMessageFromResult> t = ReceiveMessageFromAsync(new Memory<byte>(buffer, offset, size), socketFlags, remoteEP).AsTask();
+            if (t.IsCompletedSuccessfully)
```
Comment (geoffkizer) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:2410]: I don't understand why we need to do this. We should at least add a comment explaining this. (Similarly below)

**Comment by geoffkizer** [comment]
I assume all the deleted ArgumentValidationTests got migrated to tests in the appropriate test file?

Seems like it would be nice to do the same for other tests in ArgumentValidationTests.... not for this PR of course.

**Review by geoffkizer — CHANGES_REQUESTED** [review]
Just the one issue re remoteEP handling in Begin below... this is probably just a matter of adding a comment, but I'd like to understand it before we merge.

**Comment by antonfirsov** [comment]
@geoffkizer added some comments. Here is test code checking the mentioned historical behavior, it would fail without the `if (t.IsCompletedSuccessfully)` stuff:
https://github.com/dotnet/runtime/blob/fa5429326ec38b93dd6fa10e0dc3363ea36153be/src/libraries/System.Net.Sockets/tests/FunctionalTests/ReceiveMessageFrom.cs#L317-L327

It wouldn't make sense to pass `remoteEP` by `ref` if we wouldn't update it in case of synchronous completions. Note that this is the vast majority of cases (literally 100% on Windows).

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs [hunk 3]
```diff
@@ -2353,113 +2310,14 @@ public IAsyncResult BeginSendTo(byte[] buffer, int offset, int size, SocketFlags
...
-            asyncResult.StartPostingAsyncOp(false);
-
-            // Post the send.
-            DoBeginSendTo(buffer, offset, size, socketFlags, remoteEP, socketAddress, asyncResult);
-
-            // Finish, possibly posting the callback.  The callback won't be posted before this point is reached.
-            asyncResult.FinishPostingAsyncOp(ref Caches.SendClosureCache);
-
-            return asyncResult;
+            Task<int> task = SendToAsync(buffer.AsMemory().Slice(offset, size), socketFlags, remoteEP).AsTask();
```
[the remark this answers is not in the document]
Comment (antonfirsov) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs]: Thanks for the suggestion! Applied everywhere.

**Comment by antonfirsov** [comment]
> I assume all the deleted ArgumentValidationTests got migrated to tests in the appropriate test file?

Yeah, this is the case, I went through existing tests carefully to make sure there is no loss in coverage.

> Seems like it would be nice to do the same for other tests in ArgumentValidationTests.... not for this PR of course.

I have a long One Note document listing Sockets.Tests tech debt (including the existence of `ArgumentValidationTests.cs`. Maybe worth to turn it into an issue?

**Comment by geoffkizer** [comment]
> I have a One Note document listing Sockets.Tests debt (including the existence of ArgumentValidationTests.cs). Maybe worth to turn it into an issue?

Yeah, that seems good.

**Comment by geoffkizer** [comment]
> It wouldn't make sense to pass remoteEP by ref if we wouldn't update it in case of synchronous completions. Note that this is the vast majority of cases (literally 100% on Windows).

Yeah, this behavior is pretty weird -- I think this is counter to the standard IAsyncResult pattern. Thanks for clarifying why we need to do this.

**Review by geoffkizer — APPROVED** [review]

**Comment by geoffkizer** [comment]
A couple of the DualMode SendTo tests are failing, can you look at these?

**Comment by antonfirsov** [comment]
@geoffkizer it's the same test failing everywhere (I was lazy to run outerloop locally).

The cause is that `Socket_BeginSendToV4IPEndPointToV4Host_Throws` expects the method to fail with a synchronous `SocketException` in case `WSASendTo` returns an error, but our established pattern of wrapping Tasks does not work like this: when a faulted task is returned, we don't throw immediately, but convert it to an `IAsyncResult` that will throw in the corresponding `End***` call instead.

**Comment by antonfirsov** [comment]
@geoffkizer I pushed a workaround for the test failure (b3e035e55645d15bb8af0f55dd5b255c80d12cfa). Do you think the PR is OK to merge in it's current form?

**Comment by geoffkizer** [comment]
Yep, LGTM

**Comment by antonfirsov** [comment]
Outer loop test failures are unrelated: #48049, #41606, #1724, #43877

**Pull request merged**